### PR TITLE
re-open `Statix.Conn` port on closure

### PR DESCRIPTION
**Pull request by superhawk610**
~This PR adds `Statix.Supervisor`, a GenServer responsible for managing StatsD connections and reopening when a port closure is detected. To accomplish this, it introduces a new `managed` mode for connections, where port closures will `raise`.~

The previous approach isn't viable since it relies on `transmit` raising, which means that any process that calls `MyApp.Statix.increment` and similar functions that rely on `transmit` will raise if the port's closed, which isn't acceptable. This new approach simply provides the `Statix.Conn` with the name that it's registered under, so when it notices a port closure it can re-open a new port and register it under the same name.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@
 /cover
 erl_crash.dump
 *.ez
+.elixir_ls
```
Comment (Ch4s3) [inline .gitignore]: You probably want this in you global `.gitignore`

File: lib/statix.ex [hunk 2]
```diff
@@ -354,16 +354,20 @@ defmodule Statix do
...
   @doc false
   def new(module, options) do
     config = get_config(module, options)
     conn = Conn.new(config.host, config.port)
     header = IO.iodata_to_binary([conn.header | config.prefix])
+    managed = Keyword.get(options, :managed, false)
 
     %__MODULE__{
-      conn: %{conn | header: header},
+      conn: %{conn | header: header, managed: managed},
```
Comment (Ch4s3) [inline lib/statix.ex]: would supervised be a better name for this?

File: lib/statix/supervisor.ex [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    # open a managed connection to the server
+    :ok = statix_mod.connect(managed: true)
+
+    {:ok, %State{mod: statix_mod, max_reconnect_attempts: max_reconnect_attempts}}
+  end
+
+  @impl GenServer
+  def handle_info({:EXIT, _port, %Statix.Conn.PortClosedError{} = _reason}, state) do
+    # cancel previous timeout, if any
+    if state.reconnect_timeout, do: :timer.cancel(state.reconnect_timeout)
```
Comment (Ch4s3) [inline lib/statix/supervisor.ex]: could this be a function?

File: lib/statix/supervisor.ex [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+      Keyword.get(opts, :max_reconnect_attempts, @default_max_reconnect_attempts)
+
+    # open a managed connection to the server
+    :ok = statix_mod.connect(managed: true)
+
+    {:ok, %State{mod: statix_mod, max_reconnect_attempts: max_reconnect_attempts}}
+  end
+
+  @impl GenServer
+  def handle_info({:EXIT, _port, %Statix.Conn.PortClosedError{} = _reason}, state) do
```
Comment (Ch4s3) [inline lib/statix/supervisor.ex]: could we test this by creating a test UDP server and shutting it down?

Comment (JustinG721) [reply lib/statix/supervisor.ex]: This should hopefully be possible using the `test_server` already available for testing in statix

File: .gitignore (hunk 1 above)
Comment (superhawk610) [reply .gitignore]: TIL you can have a global `.gitignore`.

File: lib/statix/conn.ex [hunk 5]
```diff
@@ -42,6 +42,12 @@ defmodule Statix.Conn do
         if(is_atom(sock), do: "", else: "Statix ") <>
           "#{inspect(sock)} #{type} metric \"#{key}\" lost value #{val}" <> " due to port closure"
       end)
+
+      # open a new port for subsequent tranmissions
+      # NOTE: this port will be linked to the process that
+      # called `transmit`, so perhaps this should be handled
+      # by sending a message to a GenServer instead?
+      open(conn)
```
Comment (superhawk610) [inline lib/statix/conn.ex:50]: If necessary, we could use `:gen_udp.controlling_process/2` to reassign the parent process.

**Comment by superhawk610** [comment]
This will only resolve Erlang's `:gen_udp` port _driver_ failures, not actual network port closures (UDP is a connectionless protocol, so there's not really a concept of dropping/restarting a UDP "connection"). I'm going to close this, since it seems that our missing stat issues are caused by Datadog's StatsD server failing to receive/transmit packets.

**Pull request closed without merging**